staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -2921,6 +2946,11 @@ type PodStatus struct {
 	// More info: https://git.k8s.io/community/contributors/design-proposals/node/resource-qos.md
 	// +optional
 	QOSClass PodQOSClass `json:"qosClass,omitempty" protobuf:"bytes,9,rep,name=qosClass"`
+
+	// List of user-initiated ephemeral containers that have been run in this pod. This list is append-only.
+	// This field is alpha-level and is only honored by servers that enable the DebugContainers feature.
+	// +optional
+	EphemeralContainers []EphemeralContainer `json:"ephemeralContainers,omitempty" protobuf:"bytes,12,rep,name=ephemeralContainers"`
```

Comment:
This should probably not reuse the EphemeralContainer type. Make a new type with just the relevant fields (typemeta is not relevant, for example. Neither is metadata). Perhaps `EphemeralContainerRecord` for the name.

Did we already litigate putting the spec of the ephemeral containers in the pod's status? I don't want to argue if it's been debated ad nauseam already, but I wouldn't expect it to show up like this.

Reply:
At first I used `EphemeralContainer` in the API, but then @liggitt suggested that the client could update `Pod` directly instead, and this made things much simpler.  I've removed typemeta and metadata from `EphemeralContainer` in my latest commit.

My goal is to have one `Container`, `ContainerStatus` and `TargetContainerName` in `PodStatus` per ephemeral container, but I've gone back and forth on how to represent them. I lack perspective and would appreciate your advice.

The way it's represented now is by bundling them all together:

```go
type EphemeralContainer struct {
    Spec Container
    Status ContainerStatus
    TargetContainerName string
}

type PodStatus struct {
   ...
   EphemeralContainers []EphemeralContainer
}
```

But perhaps having Spec/Status combined violates expectations for those familiar with the API. It would be just as easy to separate out Status and rename Spec:

```go
type EphemeralContainer struct {
    ContainerSpec Container
    TargetContainerName string
}

type PodStatus struct {
    ...
    EphemeralContainers []EphemeralContainer
    EphemeralContainerStatuses []ContainerStatus
}
```

It becomes up to the clients to match `Container` to `ContainerStatus` based on `Name`, but presumably they're doing that already. Either way, the client will be able to append to the `[]EphemeralContainer` to create a new Ephemeral Container and the kubelet will populate the statuses.

Reply:
> Given my new understanding, how about this structure?
> ...

given how the flow ended up, I like that split. two thoughts:
* `spec.ephemeralContainers` should still only be set/added to via the subresource... I don't think promoting/allowing `kubectl create -f <file with ephemeral containers>` is where we want to start with this feature.
* we should be careful about inlining v1.Container in EphemeralContainer, the usability gain is minimal and it risks field name conflicts in the future

Reply:
Whether to put the the `Container` in `PodSpec` or `PodStatus` has been the fundamental disagreement between sig-node and api reviewers that's had us deadlocked for the past year.

sig-node has been against modifying the `PodSpec` because it's not a declared state that the kubelet enforces, it's the result of an action that a user took. Ideally, the `PodSpec` would be entirely immutable to simplify the kubelet's state machine and encourage best practices.

The desired debugging workflow looks like

1. User signals the kubernetes system to create a Debug Container in a pod
2. The kubelet creates a debug container and reports back the status of the container in `PodStatus`

In order to be extensible and compatible with admission controllers, we want the creation API to use a `Container` for the admission controller to examine and have `Container` used by the kubelet available for examination by future controllers. When we switched to passing the `Container` to the kubelet in the `PodStatus` we may have strayed too far towards a "spec in status" feel.

The kubelet does not preserve the `Container` that's reported in `PodStatus.EphemeralContainers`, it's overwritten based on the actual state of the pod (as with most things in `PodStatus`), and Ephemeral Containers can disappear. We probably wouldn't want the kubelet to modify the `PodSpec` to reflect this, though.

Would it change anything if `/ephemeralcontainers` accepted an `EphemeralContainer` instead of a `Pod` as @lavalamp suggested with the comparison to `Scale`?

Reply:
I agree they can be forbidden on post. And allowed on put if there's no sub resource. (email response to @liggitt that github attached in the wrong place)

@verb unfortunately I don't find that persuasive at all. And I feel really terrible about the hazing ritual we've collectively put you through on this. So I will put some responses below, but also attempt to escalate this.

> We probably wouldn't want the kubelet to modify the PodSpec to reflect this, though.

Not my first choice but not the end of the world. Controllers do in fact mutate specs, though usually of other resources.

> Would it change anything if /ephemeralcontainers accepted an EphemeralContainer instead of a Pod as @lavalamp suggested with the comparison to Scale?

I would still say these container specs belong in the pod spec. (also it doesn't make sense to me that you POST a single container to this endpoint when you can have multiple. What does the system do if you do a GET on this endpoint?)

> When we switched to passing the Container to the kubelet in the PodStatus we may have strayed too far towards a "spec in status" feel.

It's not about how things "feel". "status" has the property that it can be *reproduced* by *observation*. You can't reproduce a ContainerSpec by looking at a container in operation. So a container spec is never status.

The opposite is also true: you can never *reproduce* an observation that a user has asked for an ephemeral container. Therefore that can't be in status.

> sig-node has been against modifying the PodSpec because it's not a declared state that the kubelet enforces, it's the result of an action that a user took.

Unsurprisingly, I disagree. It can be modeled both ways. Imperative and declarative are duals. The question isn't "did the user take this action" (all updates are actions from someone) it is "how do we want to represent the state of the world". The state of the world is that a temporary container is to be attached to the pod.

Doing otherwise feels to me like letting an implementation detail drive an API.

> Ideally, the PodSpec would be entirely immutable to simplify the kubelet's state machine and encourage best practices.

This is not a realistic goal IMO. Image is already mutable, and I think it's a given that resource request/limit will eventually be mutable.

Reply:
> I would still say these container specs belong in the pod spec.

I find the reasoning behind storing what the user specified in spec.ephemeralContainers compelling.  That doesn't preclude modification being done via the subresource, especially if there are specific rules around when in the pod's lifecycle it can be done, and the types of updates allowed (e.g. you cannot remove a container from spec.ephemeralContainers... it stands as a visible record that pod had an ephemeral container run on it)

> also it doesn't make sense to me that you POST a single container to this endpoint when you can have multiple. What does the system do if you do a GET on this endpoint?

this has a strong parallel to the pods/binding subresource, which:
* is the exclusive mechanism by which the spec.nodeName field of an existing pod can be set
* only supports the Create operation
* translates that into a narrowly scoped mutation of the pod spec
* is only allowed in specific cases (cannot bind an already bound pod, cannot unbind it, etc)
* assigning a pod to a node is generally reserved for a distinct set of actors (e.g. schedulers)

the ephemeral subresource:
* is the exclusive mechanism by which the spec.ephemeralContainers field of an existing pod can be added to
* only supports the Create operation
* translates that into a narrowly scoped mutation of the pod spec
* is only allowed in specific cases (cannot run an ephemeral container that already exists, cannot remove an existing ephemeral container, etc)
* running ephemeral containers is able to be reserved for a distinct set of actors

Reply:
Is there a TL;DR of the tradeoff being considered?

Some comments on status:

Status should be considered soft state and must be reconstructable from observation or acceptable to nuke. 

Users shouldn't need to ever mutate properties of status.

Reply:
We resolved the main differences, and I've updated this PR to place EphemeralContainers in the PR. There are a couple of things I'd like to revive in this thread:

1. Should updating ephemeralContainers use a subresource?
2. If we choose to use a subresource, what should be PUT there? 

@lavalamp suggested enforcing the "only allowed for updates" semantics with POST/PUT, which I like, but I think we can't do that because of the requirement to permission this differently from other pod updates.

Right now I'm posting a `Pod` to the subresource and discarding changes to anything besides `pod.Spec.EphemeralContainers` (similar to `/status`). I will look into bindings, but I doubt I'll have the perspective to develop a strong opinion here.